controllers/agentmachine_controller.go

```diff
@@ -141,51 +146,94 @@ func (r *AgentMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request
-						log.WithError(err).Error("failed to remove the Agent's ClusterDeployment ref")
-						return &ctrl.Result{}, err
-					}
-				}
-			}
+	}
+
+	// set delete hook if not present and machine not being deleted
+	annotations := machine.GetAnnotations()
+	if _, haveMachineHookAnnotation := annotations[machineDeleteHookName]; !haveMachineHookAnnotation && machine.DeletionTimestamp == nil {
+		if annotations == nil {
+			annotations = make(map[string]string)
+		}
+		annotations[machineDeleteHookName] = ""
+		machine.SetAnnotations(annotations)
+		if err := r.Update(ctx, machine); err != nil {
+			log.WithError(err).Error("failed to add machine delete hook annotation")
+			return &ctrl.Result{}, err
+		}
+		// return early here as there's no reason to check if the machine is held up on this hook as we just created it
+		return nil, nil
+	}
+
+	// return if the machine is not waiting on this hook
+	cond := conditions.Get(machine, clusterv1.PreDrainDeleteHookSucceededCondition)
+	if cond == nil || cond.Status == corev1.ConditionTrue {
+		return nil, nil
+	}
 
-			// remove our finalizer from the list and update it.
-			controllerutil.RemoveFinalizer(agentMachine, AgentMachineFinalizerName)
-			if err := r.Update(ctx, agentMachine); err != nil {
-				log.WithError(err).Errorf("failed to remove finalizer %s from resource %s %s", AgentMachineFinalizerName, agentMachine.Name, agentMachine.Namespace)
+	if agentMachine.Status.AgentRef == nil {
+		if err := r.removeMachineDeletionHookAnnotation(ctx, machine); err != nil {
+			log.WithError(err).Error("failed to remove machine delete hook annotation")
+			return &ctrl.Result{}, err
+		}
+		return nil, nil
+	}
+
+	agent, err := r.getAgent(ctx, log, agentMachine)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.WithError(err).Infof("Failed to get agent %s. assuming the agent no longer exists", agentMachine.Status.AgentRef.Name)
+			if err := r.removeMachineDeletionHookAnnotation(ctx, machine); err != nil {
+				log.WithError(err).Error("failed to remove machine delete hook annotation")
 				return &ctrl.Result{}, err
 			}
+			return nil, nil
+		} else {
+			log.WithError(err).Errorf("Failed to get agent %s", agentMachine.Status.AgentRef.Name)
+			return &ctrl.Result{}, err
 		}
-		r.Log.Info("AgentMachine is ready for deletion")
+	}
+
+	if funk.Contains(agent.ObjectMeta.Labels, AgentMachineRefLabelKey) || agent.Spec.ClusterDeploymentName != nil {
+		r.Log.Info("Removing ClusterDeployment ref to unbind Agent")
+		delete(agent.ObjectMeta.Labels, AgentMachineRefLabelKey)
+		agent.Spec.ClusterDeploymentName = nil
+		if err := r.Update(ctx, agent); err != nil {
+			log.WithError(err).Error("failed to remove the Agent's ClusterDeployment ref")
+			return &ctrl.Result{}, err
+		}
+	}
 
-		return &ctrl.Result{}, nil
+	// definitely missed this update in the error case. Probably need to requeue?
+	removeHookStates := []string{
+		aimodels.HostStatusReclaimingRebooting,        // reclaim succeeded
```

Comment:
This status isn't going to work.

In my live test we hit this status and removed the hook before the agent actually got the command to reboot.
So the pod got drained and the node was removed, but it never rebooted into discovery.

We'll have to put some list of likely states here to make sure we don't miss anything.